Api phase9 with fixlogs

lrettig opened the pull request:
## Motivation
Same as #2071 but includes a fix for a logging issue (#2092) as well. Want to try running the system tests on this to see if it works around the issue I'm running into in #2071.

lrettig committed:
Implement MeshService.CurrentEpoch

Pass layersPerEpoch into the MeshService

lrettig committed:
Implement and add tests for 3 more endpoints

lrettig committed:
Slightly refactor tests

lrettig committed:
Move max tx per block into config

lrettig committed:
Implement MaxTransactionsPerSecond

and adds test for it

lrettig committed:
WIP implementing LayersQuery

Still need to wrangle some more data

lrettig committed:
Complete, untested implementation of LayersQuery

lrettig committed:
Trivial version of test of new endpoint passing

lrettig committed:
Implement more comprehensive test

Test all return data, test is passing
Fixed a couple of implementation bugs spotted while testing

First E2E working version of LayersQuery

lrettig committed:
Make lint happy

And add some fun comments

lrettig committed:
Handling and testing of errors

Add much more comprehensive testing of errors

Simplify how we handle errors involving missing layers for now

lrettig committed:
Add test for missing inputs

lrettig committed:
WIP: implement AccountMeshDataQuery

lrettig committed:
Add mock mempool

lrettig committed:
Overhaul tests

Clean up considerably, add a bunch more tests for AccountData

lrettig committed:
Run grpc server on localhost for test

This prevents obnoxious macOS popup

lrettig committed:
Pass txpool to new MeshService

lrettig committed:
Make lint happy

lrettig committed:
Basic E2E support for first stream

Includes a basic test

lrettig committed:
Add activations stream and test for same

lrettig committed:
WIP: sending events

service does not read filter yet

lrettig committed:
Properly handle filter in AccountMeshDataStream

lrettig committed:
Make tests much more robust

Add lots more tests for AccountMeshDataStream

Streamline the comprehensive test with another waitgroup

Fix a test issue where order of ATXs was changing due to their being
stored in a map

lrettig committed:
WIP cleaning up comprehensive stream test

lrettig committed:
Lots of cleanup

Broke out complex concurrency test that was working before, but stopped
working after sharing some data structures with the other tests I added.
It's working again as its own standalone test.

lrettig committed:
More test cleanup

- Re-add test removed by accident
- Remove test I meant to remove
- Remove commented-out test for closing stream from client side - there
isn't really any way to test this, and it doesn't seem important to test

lrettig committed:
Streamline all events reporting

Have everything flow through new streamer interface

lrettig committed:
Basic LayerStream implementation

Adds a basic test, which is passing

lrettig committed:
Add todo note about stream context

lrettig committed:
Minor test cleanup

lrettig committed:
Fully validate received layer from LayerStream

Catch an uncaught error in test
Tests for MeshService are finished

lrettig committed:
Move test function

From mesh to types, where it belongs, to avoid a dependency cycle

lrettig committed:
Rename streamer -> reporter

The name is a bit clearer and reflects its general purpose better
Add a non-API related test for the new reporting interface
Make event reporting non-blocking

lrettig committed:
Make linter happy

lrettig committed:
Very minor cleanup

Made a couple of TODO items clearer

lrettig committed:
Subsume pubsub init into new reporter

lrettig committed:
Add config for all services

Also adds a stub implementation of the GlobalStateService

lrettig committed:
Finish GlobalStateService mocks

Pass all enable service flags into JSON server
Update some description comments

lrettig committed:
Finish TxService, SmesherService stubs

lrettig committed:
Some minor cleanup

Remove err retval from Cleanup since it's unused

lrettig committed:
Add error stream

For now just hook into logs to catch error/panic

lrettig committed:
Set up Status channel infrastructure

Rename a bunch of things

lrettig committed:
Subscribe to all relevant status events

lrettig committed:
Hook into zapcore logging to report errors

This prevents a dependency loop from the log package

Make all reporter channel writes nonblocking

Add a basic test for reporting errors

lrettig committed:
Add basic test for ReportNodeStatus

lrettig committed:
WIP writing a node test for the new streams

Both streams are working. Clean up the test to check the contents of the
received messages.

lrettig committed:
Bugfix: reporter must subscribe in a goroutine

lrettig committed:
Cleanup comments

lrettig committed:
Improve NodeService stream tests

lrettig committed:
Add cmdline flag for tcp interface

This, in combination with --acquire-port=false, prevents the obnoxious
warning popup in macOS.

lrettig committed:
Add cmdline flag for grpc server interface

Make this fully specifiable. In combination with previous commit, this
fully works around obnoxious macOS warning popup.

Include this in node test.

lrettig committed:
Add channel to notify that app is started

Used in tests rather than time.Sleep, which is messy. Ideally we'd have
a full set of notification channels, one per service, so that tests
could subscribe to the services they're interested in, but this works
well enough as a stopgap for testing.

lrettig committed:
Improve test

Improve concurrency handling, check more details

lrettig committed:
More sophisticated handling of errors

Make the error channel in reporter buffered, and add a comment
explaining why

Test panicking in the test. Make sure each error is received in the
expected order.

lrettig committed:
Get state root hash for all layers

lrettig committed:
Update to latest api build

lrettig committed:
Remove two unused interface methods

lrettig committed:
Implement GlobalStateHash

Includes basic test

lrettig committed:
Implement and test Account endpoint

Includes test

Clean up method param names from API interface

lrettig committed:
Implement AccountMeshDataQuery MaxResults, Offset

lrettig committed:
Minor cleanup of filters in MeshService

lrettig committed:
WIP: partially implemented AccountDataQuery

There is currently no way to get smesher for each reward

lrettig committed:
Update comment

Add link to issue

lrettig committed:
Bugfix

spotter: @antonlerner

lrettig committed:
Handle some lingering TODO items

Fix small issues discussed with Anton in phase7 review
Sort out layer statuses for LayersQuery and LayerStream
Report layer data more often from Mesh
Encapsulate layer status inside new layers reported using the Reporter

lrettig committed:
Update txProcessor interface in test

To match changes several commits ago

lrettig committed:
Update and clean up tests

To match latest code changes.
Add some explanatory comments.
Fix a bug in Reporter.

lrettig committed:
Finish implementing AccountDataQuery

Adds basic test

lrettig committed:
Mark SmesherDataQuery as unimplemented

for now - we are waiting on #2068

lrettig committed:
WIP implementing AccountDataStream

lrettig committed:
Finish AccountDataStream

Add test
Add more streams to Reporter

lrettig committed:
Bugfix in Reporter

Spotter: @AndrewAR2

lrettig committed:
Update modules after rebase

lrettig committed:
Add blocking option to Reporter

Use it to remove time.Sleep from test

lrettig committed:
Update to latest api module version

Renames a few things

lrettig committed:
Update api module again

lrettig committed:
Rename to match latest API module

lrettig committed:
Finish GlobalStateService, add basic test

lrettig committed:
Add comprehensive test for same

lrettig committed:
Revert "Remove two unused interface methods"

This reverts commit a6a6b5141d58d520bd693683cf3cbce6bae655d2.

lrettig committed:
Update to latest API module

lrettig committed:
Rename things to match latest API build

lrettig committed:
Finish alpha implementation of SmesherService

Lots of endpoints still to be filled in

lrettig committed:
Add basic SmesherService tests and stubs

lrettig committed:
Properly handle input errors

lrettig committed:
Add more tests for unimplemented endpoints

lrettig committed:
Minor cleanup

Cleanup some unused properties in several of the services
Clean up some comments

lrettig committed:
Initial complete implementation of TxService

Adds a test for SubmitTransaction only so far
Clean up the test code a bit
Modify Reporter transaction channel to include tx state
Report invalidated txs from tx pool

lrettig committed:
More comprehensive tests

lrettig committed:
Finish adding TxService tests

lrettig committed:
Add another more robust stream test

lrettig committed:
Update mocked object to match interface change

See c373c24bc94ccf339586d5c6e685e9efa7a3482b

lrettig committed:
Do not allow reporter to be reinitialized

Unless closed first

lrettig committed:
Update service arguments

To match cleanup of their constructors

lrettig committed:
Add AlwaysListen config

Forces syncer to always listen to gossip, useful for testing

lrettig committed:
Add E2E test of tx service

lrettig committed:
Add more bad input tests

lrettig committed:
Tidy modules after rebase

lrettig committed:
Make status reporting much simpler

Don't convey any data as part of the reporting. Let the listener pull
down all of the data. This makes everything simpler and more
streamlined. Update tests accordingly.

lrettig committed:
Add issue links for TODO items

lrettig committed:
Make linter happy

lrettig committed:
Make linter great again

lrettig committed:
Fix test

lrettig committed:
Fix test for CI and Docker

Ignore an error that only happens when there's an existing database

lrettig committed:
Small tweaks to fix tests

lrettig committed:
Skip two defunct tests for now

They will be re-enabled in phase9. The required fixes are too complex to
backport.

lrettig committed:
Register all services to gateway server

lrettig committed:
Register all services before starting grpc server

This mostly worked before due to luck, but this is the order things
should happen in. Make sure they happen in the right order.

lrettig committed:
Make tests less fragile

Make sure global memory is cleared before the NodeService test, which
watches the error stream. Add a sleep statement since we have no way of
ensuring th service is actually listening and this keeps causing issues
in docker/travis.

lrettig committed:
Try fixing the leaking another way

The previous fix did not work because the log itself is a global
singleton. Try another workaround until we have a more robust solution.

lrettig committed:
Make lint happy

Clean up tests
Rename a GRPC method to make lint happy

lrettig committed:
Fix merge conflicts, update API version

Update naming to match latest API build
Finish resolving merge conflicts from rebase on phase8

lrettig committed:
Always close reporter before reinit in tests

Should prevent error about already being initialized

lrettig committed:
Re-enable and fix failing tests

Log the number of registered services from grpc server

lrettig committed:
Bugfix: run http server in a goroutine

Same as grpc server, since this blocks

lrettig committed:
Bugfix: make sure context isn't closed too early

Run the entire surrounding method in a goroutine

lrettig committed:
Minor improvements to tests

Make sure the event reporter is closed between tests

lrettig committed:
Make test less flakey

Adding a brief sleep (womp womp) seems to prevent sporadic test failures
due to listener not listening yet

lrettig committed:
Small fix to test

lrettig committed:
Make thread-safe

lrettig committed:
Remove an unnecessary check

We don't perform this check elsewhere. It shouldn't be necessary. Per
@antonlerner in review.

lrettig committed:
Fix test

lrettig committed:
Improve concurrency in test

lrettig committed:
Better isolate tests

The global Context was causing one test to cancel another. Protect the
tests with a setup method that isolates them, and have each test use a
different port to isolate them even more.

lrettig committed:
Remove unused args from tx service

lrettig committed:
Some minor cleanup

lrettig committed:
Small tweaks to improve logging

Print full error message from mesh
Pretty-print layer info in reporter logs

lrettig committed:
Remove a redundant report

lrettig committed:
WIP troubleshoot LayerStream issue

See #2084

lrettig committed:
Fix error about missing layer state root

This is expected. Update log message accordingly.

lrettig committed:
Fix debug message

lrettig committed:
Upgrade to latest zap module

lrettig committed:
Fix log levels issue

Simplify the wrapped Log type
Set level of derived logger correctly, and warn if the level is too low

lrettig committed:
Clean up some misc log statements

lrettig committed:
Add link to issue for TODO

lrettig committed:
Fix lingering merge conflict

Not sure how I missed this!

lrettig committed:
Bugfix: correctly calculate CommitmentSize

This refers to the amount of space being committed to, not to the size of the ATX itself!

lrettig committed:
Make linter happy, fix a test

Don't try to predict in the test what happens after the app closes and
the stream closes, this appears to be nondeterministic

lrettig committed:
Fix another test

lrettig committed:
Fix test

lrettig committed:
Bugfix, add debug output

Catch empty input before it causes a null pointer exception

lrettig committed:
Final cleanup, per review

per Anton's review
reuse error message
factor out some code to make parent method shorter

lrettig committed:
More cleanup post refactor

lrettig committed:
One more for the refactor

Fix https://github.com/spacemeshos/go-spacemesh/issues/2116

lrettig committed:
Fix more imports post-rebase

lrettig committed:
Fix tests post-rebase

Check for nil pointers

lrettig committed:
Merge branch 'fix-log-levels3' into api-phase9-with-fixlogs

lrettig committed:
fix merge issue

lrettig committed:
Make endpoint function description clearer

lrettig committed:
Merge branch 'api-phase9' into api-phase9-with-fixlogs

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

antonlerner commented on log/zap.go:
```diff
@@ -129,57 +127,27 @@ func (l Log) With() FieldLogger {
-		return ce
-	}
-	return ce.AddCore(e, c.Core)
+	lgr := l.logger.Named(fmt.Sprintf("%-13s", prefix))
```
this completely eliminates dynamic level debugging

lrettig replied:
Yes, that's the idea - see #2090 and #2092 for background. I think we're abusing the log levels here, using them in ways they're not intended to be used, which causes downstream problems. I'll check if I can split this into two, and include just a one- or two-line fix for now, and leave the rest as a separate PR.

antonlerner commented on cmd/node/node.go:
```diff
@@ -445,10 +446,8 @@ func (app *SpacemeshApp) initServices(nodeID types.NodeID,
-	lg := log.NewDefault(name).WithFields(nodeID)
-
+	// This base logger must be debug level so that other, derived loggers are not a lower level.
+	lg := log.NewWithLevel(nodeID.ShortString(), zap.NewAtomicLevelAt(zapcore.DebugLevel)).WithFields(nodeID)
```
I think this is the only change needed - to have the default logger run with debug level.

lrettig replied:
I tested this (https://github.com/spacemeshos/go-spacemesh/pull/2071/commits/d1c68d28339c8da7070d5ed066fe71f917afa7c3), it didn't work. The issue is deeper than this. I explained why in #2090. I don't think we can make a change this superficial here.

lrettig commented:
bors try

lrettig commented:
bors ping

lrettig commented:
bors try

lrettig commented:
> Same as #2071 but includes a fix for a logging issue (#2092) as well. Want to try running the system tests on this to see if it works around the issue I'm running into in #2071.

Confirmed that the failure in #2071 is due to the logging issue, and that the fix in this branch fixes it.

lrettig committed:
Merge branch 'develop' into api-phase9-with-fixlogs

lrettig committed:
Do a better job of sending test transactions

Fail fast when generating and sending transactions to new test accounts
fails
Pause after sending each tx to give the network a chance to gossip the
tx before sending another from the same source account

lrettig committed:
Merge remote-tracking branch 'origin/test-submit-tx-fix' into api-phase9-with-fixlogs

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
bors try

lrettig commented:
bors try

lrettig committed:
Merge branch 'develop' into api-phase9-with-fixlogs

lrettig commented:
This PR was not meant to be merged, it was just a test to see if adding the logging fixes from #2092 would cause the tests to pass. They're passing. Closing this now, I'll follow up in https://github.com/spacemeshos/go-spacemesh/pull/2071#issuecomment-681021706.
